_posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md

```diff
@@ -0,0 +1,54 @@
+
+At first sight, it seems like a relatively simple task to make sure that this math is correct. It turns out that handling edge cases near the min and max durations or when performing operations between very large and very small durations requires special attention. Kani has helped fix at least eight different categories of bugs in a single pull request: <https://github.com/nyx-space/hifitime/pull/192>. Most of these were bugs near the boundaries of a Duration definition: around the zero, maximum, and minimum durations. But many of the bugs were on important and common operations: partial equality, negation, addition, and subtraction operations. These bugs weren't due to lax testing: there are over 74 integration tests with plenty of checks within each.
+
+One of the great features of Kani is that it performs what is known as symbolic execution of programs, where inputs are modelled as symbolic variables covering whole ranges of values at once. All program behaviors possible under these inputs are analyzed for defects like arithmetic overflows or underflows, signed conversion overflow or underflow, etc.. If a defect is possible for some values of the inputs, Kani will generate a counter example trace with concrete values triggering the defect.
+
+Thanks to how Kani analyzes a program, tests can either be bounded or not. A bounded test includes an assertion: execute a set of instructions and then check something. This is a typical test case.
+
+Kani can also have unbounded tests where there is no explicit condition to check. Instead, only the successive operations of a function call are executed, and each are tested by Kani for failure cases by analyzing the inputs and finding cases where inputs will lead to runtime errors like overflows. This approach is how most of the bugs in hifitime have been found.
+
+Unbounded tests effectively ensure that sanity of the operations in a given function call. Bounded tests provide the same while also checking equalities. If either of these tests fail, Kani can provide a test failure report outlining the sequence of operations, and the binary representation of each intermediate operation, to help the developer gain an understanding of why their implementation is incorrect.
+
+The overhead to implement tests in Kani is very low, and the benefits are immense. Basically, write a Kani verification like a unit test, run the model verifier, and you've formally verified this part of the code. Amazing!
```

Comment:
I think it's really important here to highlight how assumptions are important as well. The verification result is as strong as your assumptions.

Reply:
cc. @zhassan-aws @tautschnig @rahulku